common.go

```diff
@@ -641,6 +641,18 @@ type Config struct {
 	// This value has no meaning for the client.
 	GetDelegatedCredential func(*ClientHelloInfo, uint16) ([]byte, crypto.PrivateKey, error)
 
+	// ClientESNIKeys enables SNI encryption by the client using the
+	// provided server keys, it has no effect on a server. If the value is
+	// nil, then SNI encryption is not enabled.
+	//
+	// See https://tools.ietf.org/html/draft-ietf-tls-esni-01
+	ClientESNIKeys *ESNIKeys
```

Comment:
This doesn't fit the pattern of a config being reused across connections since the ESNI keys are server specific. I can't think of a great way to do it, but one way would be to have a flag here to enable support, then DialWithDialer would use the resolver passed in to look up the text record.